File: utils/number.ts
export function parseNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }

  if (typeof value !== "string") {
    return null;
  }

  const normalized_value = value
    .trim()
    .replaceAll(/[$,%\s]/g, "")
    .replaceAll(",", "");

  if (!normalized_value) {
    return null;
  }

  const parsed_value = Number(normalized_value);
  return Number.isFinite(parsed_value) ? parsed_value : null;
}

export function roundNumber(value: number, decimals = 2): number {
  const factor = 10 ** decimals;
  return Math.round((value + Number.EPSILON) * factor) / factor;
}

export function formatNumber(value: number): string {
  if (Number.isInteger(value)) {
    return value.toString();
  }

  return value.toString();
}

File: src/shared/utils/number.rs
pub fn parse_number(value: &str) -> Option<f64> {
    let normalized = value
        .trim()
        .replace(['$', '%', ' ', '\t', '\n', '\r'], "")
        .replace(',', "");

    if normalized.is_empty() {
        return None;
    }

    normalized
        .parse::<f64>()
        .ok()
        .filter(|value| value.is_finite())
}

pub fn round_number(value: f64, decimals: u32) -> f64 {
    let factor = 10_f64.powi(decimals as i32);
    ((value + f64::EPSILON) * factor).round() / factor
}

pub fn format_number(value: f64) -> String {
    if value.fract() == 0.0 {
        format!("{:.0}", value)
    } else {
        let mut text = value.to_string();
        if text.contains('.') {
            while text.ends_with('0') && text.contains('.') {
                text.pop();
            }
            if text.ends_with('.') {
                text.pop();
            }
        }
        text
    }
}
